src/engine/Sort.cpp

```diff
@@ -54,121 +58,148 @@ std::string Sort::getDescriptor() const {
-    return computeResultExternal(subRes, requestLaziness);
-  } else {
-    const auto& subTable = subRes->idTable();
-    getExecutionContext()->getSortPerformanceEstimator().throwIfEstimateTooLong(
-        subTable.numRows(), subTable.numColumns(), deadline_, "Sort operation");
-    return computeResultInMemory(subRes);
+  size_t numColumns = subtree_->getResultWidth();
+  // Maximum number of rows that can be sorted in memory.
+  size_t maxNumRowsToBeSortedInMemory =
+      getRuntimeParameter<&RuntimeParameters::sortInMemoryThreshold_>()
+          .getBytes() /
+      (numColumns * sizeof(Id));
+
+  // Always request lazy input to avoid premature materialization.
+  std::shared_ptr<const Result> subRes = subtree_->getResult(true);
+
+  // For fully materialized input, we know the size upfront.
+  if (subRes->isFullyMaterialized()) {
+    const IdTable& table = subRes->idTable();
+    if (table.numRows() <= maxNumRowsToBeSortedInMemory) {
+      return computeResultInMemory(table.clone(),
+                                   subRes->getSharedLocalVocab());
+    } else {
+      std::vector<IdTable> blocks;
+      blocks.push_back(table.clone());
+      std::vector<Result::IdTableVocabPair> noRemainingBlocks;
+      return computeResultExternal(
+          std::move(blocks), subRes->getCopyOfLocalVocab(),
+          noRemainingBlocks.begin(), noRemainingBlocks.end(), requestLaziness);
+    }
+  }
+
+  // For lazy input, collect blocks until we exceed the threshold. Note that we
+  // may exceed the threshold by the size of one block.
+  std::vector<IdTable> collectedBlocks;
+  LocalVocab mergedLocalVocab;
+  size_t totalRows = 0;
+  auto idTables = subRes->idTables();
+  auto it = idTables.begin();
+  while (it != idTables.end() && totalRows <= maxNumRowsToBeSortedInMemory) {
+    checkCancellation();
+    auto& idTableAndLocalVocab = *it;
+    totalRows += idTableAndLocalVocab.idTable_.numRows();
+    collectedBlocks.push_back(std::move(idTableAndLocalVocab.idTable_));
+    mergedLocalVocab.mergeWith(idTableAndLocalVocab.localVocab_);
+    ++it;
+  }
+
+  // If we exceeded the threshold (by one block), use external sort.
+  if (totalRows > maxNumRowsToBeSortedInMemory) {
+    return computeResultExternal(std::move(collectedBlocks),
+                                 std::move(mergedLocalVocab), std::move(it),
+                                 idTables.end(), requestLaziness);
   }
+
+  // Stayed under threshold: concatenate and sort in-memory.
+  IdTable combined{numColumns, allocator()};
+  combined.reserve(totalRows);
+  for (auto& block : collectedBlocks) {
+    combined.insertAtEnd(std::move(block));
+  }
+  return computeResultInMemory(std::move(combined),
+                               std::move(mergedLocalVocab));
 }
 
 // _____________________________________________________________________________
-Result Sort::computeResultInMemory(std::shared_ptr<const Result> subRes) const {
+template <typename LocalVocabT>
```

Comment:
I don't think you need to template the type of the local vocab here. The performance difference of calling clone vs. incrementing the ref count of a shared pointer is small compared to the time it takes to sort everything.